File: CSharp/DP.UnitTest/UtMediator.cs
using System;
using System.Diagnostics;
using DP.Domain.Samples.Mediator;
using DP.UnitTest.Utility;
using Xunit;
using Xunit.Abstractions;

namespace DP.UnitTest
{
    public class UtMediator
    {
        private readonly ITestOutputHelper _output;

        public UtMediator(ITestOutputHelper output)
        {
            this._output = output;
            System.Diagnostics.Trace.Listeners.Add(new XunitTraceListener(this._output));
        }

        [Fact]
        public void TestWeightScore()
        {
            var weightsForOption = new decimal[] { 0.2M, 0.5M, 0.8M };
            var weightsForCredit = new decimal[] { 0.1M, 0.3M, 0.2M };
            var weightsForLoan = new decimal[] { 0.6M, 0.2M, 0 };

            decimal expectedOptionScore = (new OptionColleague()).Score() * weightsForOption[0] +
                        (new CreditColleague()).Score() * weightsForOption[1] +
                        (new LoanColleague()).Score() * weightsForOption[2];

            decimal expectedCreditScore = (new OptionColleague()).Score() * weightsForCredit[0] +
                       (new CreditColleague()).Score() * weightsForCredit[1] +
                       (new LoanColleague()).Score() * weightsForCredit[2];

            decimal expectedLoanScore = (new OptionColleague()).Score() * weightsForLoan[0] +
                       (new CreditColleague()).Score() * weightsForLoan[1] +
                       (new LoanColleague()).Score() * weightsForLoan[2];

            #region Mediator for Option
            IMediator mediatorForOption = new MediatorWeight(weightsForOption[0], weightsForOption[1], weightsForOption[2]);
            IColleague option = new OptionColleague(mediatorForOption);
            //Score!
            decimal actualOptionScore = option.Mediator.Score();
            Trace.WriteLine($"{option.Prod} 權重計分結果={actualOptionScore.ToString()}");

            Assert.Equal(expectedOptionScore, actualOptionScore);
            #endregion        

            #region Mediator for Credit
            IMediator mediatorForCredit = new MediatorWeight(weightsForCredit[0], weightsForCredit[1], weightsForCredit[2]);
            IColleague credit = new CreditColleague(mediatorForCredit);
            //Score!
            decimal actualCreditScore = credit.Mediator.Score();
            Trace.WriteLine($"{credit.Prod} 權重計分結果={actualCreditScore.ToString()}");

            Assert.Equal(expectedCreditScore, actualCreditScore);            
            #endregion

            #region Mediator for Loan
            IMediator mediatorForLoan = new MediatorWeight(weightsForLoan[0], weightsForLoan[1], weightsForLoan[2]);
            IColleague loan = new LoanColleague(mediatorForLoan);
            //Score!
            decimal actualLoanScore = loan.Mediator.Score();
            Trace.WriteLine($"{credit.Prod} 權重計分結果={actualLoanScore.ToString()}");
            
            Assert.Equal(expectedLoanScore, actualLoanScore);
            #endregion   
        }

        [Fact]
        public void TestAverageScore()
        {
            decimal expectedLoanScore = ((new OptionColleague()).Score() +
                        (new CreditColleague()).Score() +
                        (new LoanColleague()).Score())/3;

            #region Mediator
            IMediator mediator = new MediatorAverage();
            IColleague loan = new LoanColleague(mediator);
            //Score!
            decimal actualLoanScore = loan.Mediator.Score();
            #endregion

            Trace.WriteLine($"{loan.Prod} 平均計分結果={actualLoanScore.ToString()}");

            Assert.Equal(expectedLoanScore, actualLoanScore);
        }
    }
}

File: Python/Samples/Mediator/UtMediator.py
import unittest
from Mediator import Mediator
from OptionColleague import OptionColleague
from CreditColleague import CreditColleague
from LoanColleague import LoanColleague
from MediatorWeight import MediatorWeight
from MediatorAverage import MediatorAverage


class UtMediator(unittest.TestCase):

    def test_weight_score(self):

        weightsForOption = [0.2, 0.5, 0.8]
        weightsForCredit = [0.1, 0.3, 0.2]
        weightsForLoan = [0.6, 0.2, 0]

        expectedOptionScore = (OptionColleague()).score() * weightsForOption[0] + (CreditColleague()).score() * weightsForOption[1] + (LoanColleague()).score() * weightsForOption[2]

        expectedCreditScore = (OptionColleague()).score() * weightsForCredit[0] + (CreditColleague()).score() * weightsForCredit[1] + (LoanColleague()).score() * weightsForCredit[2]

        expectedLoanScore = (OptionColleague()).score() * weightsForLoan[0] + (CreditColleague()).score() * weightsForLoan[1] + (LoanColleague()).score() * weightsForLoan[2]

        # Mediator for Option
        mediatorForOption = MediatorWeight(
            weightsForOption[0], weightsForOption[1], weightsForOption[2])
        option = OptionColleague()
        option.mediator = mediatorForOption
        actualOptionScore = option.mediator.score()  # Score!
        print("{0} 權重計分結果={1}".format(option.prod, actualOptionScore))

        self.assertEqual(actualOptionScore,  expectedOptionScore)

        # Mediator for Credit
        mediatorForCredit = MediatorWeight(
            weightsForCredit[0], weightsForCredit[1], weightsForCredit[2])
        option = CreditColleague()
        option.mediator = mediatorForCredit
        actualCreditScore = option.mediator.score()  # Score!
        print("{0} 權重計分結果={1}".format(option.prod, actualCreditScore))

        self.assertEqual(actualCreditScore,  expectedCreditScore)
 
        # Mediator for Loan
        mediatorForLoan = MediatorWeight(
            weightsForLoan[0], weightsForLoan[1], weightsForLoan[2])
        option = LoanColleague()
        option.mediator = mediatorForLoan
        actualLoanScore = option.mediator.score()  # Score!
        print("{0} 權重計分結果={1}".format(option.prod, actualLoanScore))

        self.assertEqual(actualLoanScore,  expectedLoanScore)

if __name__ == '__main__':
    unittest.main()
